### PR TITLE
Add Settings.StyleCop file

### PR DESCRIPTION
**Pull request by iSazonov**
## PR Summary

Related #4708.

Disable:
1. [SA1402 FileMayOnlyContainASingleType
](https://github.com/DotNetAnalyzers/StyleCopAnalyzers/blob/master/documentation/SA1402.md)
1. [SA1403 FileMayOnlyContainASingleNamespace
](https://github.com/DotNetAnalyzers/StyleCopAnalyzers/blob/master/documentation/SA1403.md) 
1. [SA1101 PrefixLocalCallsWithThis](https://github.com/DotNetAnalyzers/StyleCopAnalyzers/blob/master/documentation/SA1101.md)
1. [SA1126 PrefixCallsCorrectly](https://github.com/DotNetAnalyzers/StyleCopAnalyzers/blob/master/documentation/SA1126.md)

[stylecop.json](https://github.com/DotNetAnalyzers/StyleCopAnalyzers/blob/master/documentation/Configuration.md)

## PR Checklist

- [x] [PR has a meaningful title](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
    - Use the present tense and imperative mood when describing your changes
- [x] [Summarized changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [x] [Change is not breaking](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#making-breaking-changes)
- [x] [Make sure all `.h`, `.cpp`, `.cs`, `.ps1` and `.psm1` files have the correct copyright header](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [x] This PR is ready to merge and is not [Work in Progress](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---work-in-progress).
    - If the PR is work in progress, please add the prefix `WIP:` to the beginning of the title and remove the prefix when the PR is ready.
- **User-facing changes**
    - [x] Not Applicable
    - **OR**
    - [ ] User-facing [Documentation needed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
        - [ ] Issue filed - Issue link:
- **Testing - New and feature**
    - [x] Not Applicable or can only be tested interactively
    - **OR**
    - [ ] [Make sure you've added a new test if existing tests do not effectively test the code changed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#before-submitting)
        - [ ] [Add `[feature]` if the change is significant or affects feature tests](https://github.com/PowerShell/PowerShell/blob/master/docs/testing-guidelines/testing-guidelines.md#requesting-additional-tests-for-a-pr)

### CODE REVIEW

**Comment by markekraus** [comment]
I think for this project we need to disable [SA1200](https://github.com/DotNetAnalyzers/StyleCopAnalyzers/blob/master/documentation/SA1200.md) too. This project almost exclusively has using statements outside the namespace. 

As for whether that is a good thing or not... is debatable.

**Comment by TravisEz13** [comment]
restarted CI due to an update-help connection failure in travis-ci

**Comment by TravisEz13** [comment]
@SteveL-MSFT @adityapatwardhan @daxian-dbw I've put the list of what is being disabled in the description.  Are we all ok with this?  I prefer a file only contain a single class, but I agree that it is not what we have been doing.

**Review by TravisEz13 — COMMENTED** [review]
I prefer files only contain a single class, but I agree the code base doesn't follow this standard.  I'm fine with this if others agree.

**Comment by markekraus** [comment]
My opinion on this is that we should do what is best for an Open Source project, whatever that is. 

One problem I hear from people acclimating to the code base is that these monolith files are both intimidating and hard to follow. It is often gating low complexity changes from novice contributors.

I personally believe it is easier to track changes from larger numbers of people if those changes are spread over a larger number of files.

To both those ends, one class per file is probably better.

**Comment by iSazonov** [comment]
While IntelliSense works in VS Code I haven't difficulties with multi-class files. :-)

Our guideline says that we don't want style changes but CodeFactor fixes is style changes. Should we ignore/remove the requirement or replace with requirement to fix CodeFactor issues?

Also we have a requirement to follow current file style. This makes the code easier to read. If we follow the CodeFactor requirements we break the readability.

I suppose we should first set a top policy:
- should we fix **all** (approved/not disabled) CodeFactor issues
- should we fix problems for **specific rules** only

**Comment by iSazonov** [comment]
Fix all 186264 issues is very hude work even if we turn down some rules.

**Comment by markekraus** [comment]
@iSazonov Agreed. This problem is why I ultimately stopped investigating style enforcement, The project has guidelines which are not compatible with style enforcement (use style of surrounding code, don't make style changes), tools for style enforcement are not precise enough to limit detection to just the code added or change in the PR (makes it useless for reviews), and fixing the style globally so that rules and tools can be useful is a massive undertaking. Also, trying to fix the problems in a single area (like web cmdlets) is a strain on the reviewers and the contributor, there are not enough contributors and reviewers to tackle the problem in just one area, let alone globally.

**Comment by iSazonov** [comment]
Some issues can be easily fixed:
"" -> string.Empty
String -> string

It is up to 1-2% issues.

Do you know of other problems that can be easily fixed?

**Comment by markekraus** [comment]
* `Char` -> `char`
* "Closing Curly braces must be followed by a blank line"
* "Elements must be separated by a blank line"
* Single line comments must begin with a space

Basically, the white space rules are all easy wins and there are a ton of them.

**Comment by markekraus** [comment]
We could also properly configure "The file header must contain a copyright tag" so that it doesn't false-alarm or just disable [SA1634](https://github.com/DotNetAnalyzers/StyleCopAnalyzers/blob/master/documentation/SA1634.md).

[SA1518](https://github.com/DotNetAnalyzers/StyleCopAnalyzers/blob/master/documentation/SA1518.md) should be set to required.

These are some of the easy low risk wins, I think:

- https://www.codefactor.io/repository/github/powershell/powershell/pull/6930?category=Maintainability&groupId=259
- https://www.codefactor.io/repository/github/powershell/powershell/pull/6930?category=Maintainability&groupId=281
- https://www.codefactor.io/repository/github/powershell/powershell/pull/6930?category=Maintainability&groupId=46
- https://www.codefactor.io/repository/github/powershell/powershell/pull/6930?category=Maintainability&groupId=22
- https://www.codefactor.io/repository/github/powershell/powershell/pull/6930?category=Style&groupId=81
- https://www.codefactor.io/repository/github/powershell/powershell/pull/6930?category=Style&groupId=72
- https://www.codefactor.io/repository/github/powershell/powershell/pull/6930?category=Style&groupId=264
- https://www.codefactor.io/repository/github/powershell/powershell/pull/6930?category=Style&groupId=93
- https://www.codefactor.io/repository/github/powershell/powershell/pull/6930?category=Style&groupId=80
- https://www.codefactor.io/repository/github/powershell/powershell/pull/6930?category=Style&groupId=83
- https://www.codefactor.io/repository/github/powershell/powershell/pull/6930?category=Style&groupId=95
- https://www.codefactor.io/repository/github/powershell/powershell/pull/6930?category=Style&groupId=6
- https://www.codefactor.io/repository/github/powershell/powershell/pull/6930?category=Style&groupId=40
- https://www.codefactor.io/repository/github/powershell/powershell/pull/6930?category=Style&groupId=85
- https://www.codefactor.io/repository/github/powershell/powershell/pull/6930?category=Style&groupId=39
- https://www.codefactor.io/repository/github/powershell/powershell/pull/6930?category=Style&groupId=38
- https://www.codefactor.io/repository/github/powershell/powershell/pull/6930?category=Style&groupId=18
- https://www.codefactor.io/repository/github/powershell/powershell/pull/6930?category=Style&groupId=8
- https://www.codefactor.io/repository/github/powershell/powershell/pull/6930?category=Style&groupId=82
- https://www.codefactor.io/repository/github/powershell/powershell/pull/6930?category=Style&groupId=9
- https://www.codefactor.io/repository/github/powershell/powershell/pull/6930?category=Style&groupId=269
- https://www.codefactor.io/repository/github/powershell/powershell/pull/6930?category=Style&groupId=5

**Comment by iSazonov** [comment]
@markekraus Thanks!

SA1518 is on by default.

> the white space rules are all easy wins and there are a ton of them.

Maybe formatting tool from [.NET code formatter tool](https://github.com/dotnet/codeformatter) can help? Our [docs](https://github.com/PowerShell/PowerShell/blob/ec1106a329ed0de5681d7b9cf01a85ff5158caa6/docs/dev-process/coding-guidelines.md) says :-)
>We also run the .NET code formatter tool regularly to keep consistent formatting.

Where can get a config for the CodeFormatter?

**Comment by lzybkr** [comment]
I'd also ignore the `this.` prefix warning. More unnecessary churn if you change it, and **much** riskier to use a regex to "fix".

**Comment by SteveL-MSFT** [comment]
I'm fine with the proposed disabled settings.  In general, we should error on the side of productivity and lowering the barrier for contributions.

**Comment by iSazonov** [comment]
@lzybkr 
>  much riskier to use a regex to "fix".

I did not use a script with Regex. I did the changes in VS Code step by step selecting a pattern and then visually checking the results to exclude false changes, then commit changes for the pattern, then select next pattern.

**Comment by iSazonov** [comment]
Disable SA1101 and SA1126 - up to 25% win.

**Comment by iSazonov** [comment]
I think we should disable [SA1600 ElementsMustBeDocumented](https://github.com/DotNetAnalyzers/StyleCopAnalyzers/blob/master/documentation/SA1600.md) rule because it give strange issues like https://github.com/PowerShell/PowerShell/blob/4c925f58b80893690532c26788a913fdd35fdb70/src/System.Management.Automation/namespaces/FileSystemProvider.cs#L7344

**Comment by lzybkr** [comment]
@iSazonov - removing `this.` can change a program's behavior. Consider:

```c#
    class C {
        int value;
        void f(int v1, int value) {
            this.value = v1;
        }
    }
```

If you remove `this.` - you will be assigning to the parameter. I'm not aware of tools that will help detect this change - it's very manual. This is why I said it's risky.

**Comment by iSazonov** [comment]
In #6949 @lzybkr suggest disable [SA1131 UseReadableConditions](https://github.com/DotNetAnalyzers/StyleCopAnalyzers/blob/0298ccf2d632c19e0a72ec640f95e8c72daf00b0/documentation/SA1131.md)

**Comment by iSazonov** [comment]
I think we can merge, look how CodeFactor statistics will be changed for master branch and continue adjust the settings in follow PRs.

**Review by markekraus — APPROVED** [review]
